server/client.go

```diff
@@ -301,73 +365,202 @@ func (c *client) readLoop() {
-					}
-				}
+			// Update last activity for message delivery
+			cp.last = last
+			cp.out.fsp--
+			if budget > 0 && cp.flushOutbound() {
+				budget -= cp.out.lft
+			} else {
+				cp.flushSignal()
 			}
 			cp.mu.Unlock()
 			delete(c.pcd, cp)
 		}
-		// Check to see if we got closed, e.g. slow consumer
+
+		// Update activity, check read buffer size.
 		c.mu.Lock()
 		nc := c.nc
+
 		// Activity based on interest changes or data/msgs.
-		if c.cache.inMsgs > 0 || c.cache.subs > 0 {
+		if c.in.msgs > 0 || c.in.subs > 0 {
 			c.last = last
 		}
+
+		if n >= cap(b) {
+			c.in.srs = 0
+		} else if n < cap(b)/2 { // divide by 2 b/c we want less than what we would shrink to.
+			c.in.srs++
+		}
+
+		// Update read buffer size as/if needed.
+		if n >= cap(b) && cap(b) < maxBufSize {
+			// Grow
+			c.in.rsz = cap(b) * 2
+			b = make([]byte, c.in.rsz)
+		} else if n < cap(b) && cap(b) > minBufSize && c.in.srs > shortsToShrink {
+			// Shrink, for now don't accelerate, ping/pong will eventually sort it out.
+			c.in.rsz = cap(b) / 2
+			b = make([]byte, c.in.rsz)
+		}
 		c.mu.Unlock()
+
+		// Check to see if we got closed, e.g. slow consumer
 		if nc == nil {
 			return
 		}
+	}
+}
+
+// collapsePtoNB will place primary onto nb buffer as needed in prep for WriteTo.
+// This will return a copy on purpose.
+func (c *client) collapsePtoNB() net.Buffers {
+	if c.out.p != nil {
+		p := c.out.p
+		c.out.p = nil
+		return append(c.out.nb, p)
+	}
+	return c.out.nb
+}
+
+// This will handle the fixup needed on a partial write.
+// Assume pending has been already calculated correctly.
+func (c *client) handlePartialWrite(pnb net.Buffers) {
+	nb := c.collapsePtoNB()
+	// The partial needs to be first, so append nb to pnb
+	c.out.nb = append(pnb, nb...)
+}
+
+// flushOutbound will flush outbound buffer to a client.
+// Will return if data was attempted to be written.
+// Lock must be held
+func (c *client) flushOutbound() bool {
+	if c.flags.isSet(flushOutbound) {
+		return false
+	}
+	c.flags.set(flushOutbound)
+	defer c.flags.clear(flushOutbound)
+
+	// Check for nothing to do.
+	if c.nc == nil || c.srv == nil || c.out.pb == 0 {
+		return true // true because no need to queue a signal.
+	}
+
+	// Snapshot opts
+	srv := c.srv
+
+	// Place primary on nb, assign primary to secondary, nil out nb and secondary.
+	nb := c.collapsePtoNB()
+	c.out.p, c.out.nb, c.out.s = c.out.s, nil, nil
+
+	// For selecting primary replacement.
+	cnb := nb
+
+	// In case it goes away after releasing the lock.
+	nc := c.nc
+	attempted := c.out.pb
+	apm := c.out.pm
+
+	// Do NOT hold lock during actual IO
+	c.mu.Unlock()
+
+	// flush here
+	now := time.Now()
+	// FIXME(dlc) - writev will do multiple IOs past 1024 on
+	// most platforms, need to account for that with deadline?
+	nc.SetWriteDeadline(now.Add(c.out.wdl))
+	// Actual write to the socket.
+	n, err := nb.WriteTo(nc)
+	nc.SetWriteDeadline(time.Time{})
+	lft := time.Since(now)
+
+	// Re-acquire client lock
+	c.mu.Lock()
+
```

Comment:
Should we check if connection was closed, and if so simply return true here? Not sure it makes sense to continue after this point if it has been closed. Also, we may end-up logging a flush error if the write returned an error due to socket close.

Reply:
We check for err != nil right below here. If connection is properly closed the err should be nil, and the code below just does accounting. But we could check for that condition if you think its critical.

Reply:
It is possible that connection is closed before we grab the lock but after the `nb.WriteTo()` so err would be nil. Again, as long as the remaining code is safe if the connection is closed I am fine.